### ISSUE TITLE
머딥 - 17.3d (미분 경사하강법 적용)

### ISSUE THREAD

**Issue by vmtmxmf5** [vmtmxmf5/Python-Basics#52]
# 목차
1. 미분(numerical gradient)
2. 최적화(gradient descent) 개념
3. Loss function 과 최적화 결합

### Summary
[1] 코드로 구현한 미분이란?
> f(x+h) - f(x-h) / 2h

[2] 코드로 구현한 경사하강법이란?
> 좌표 - (학습률*미분한 좌표)

[3] 손실함수를 **미분**하려면 **lambda 함수**를 씌워줘야 한다
> numerical_gradient(f(loss(W@X)), net.W)
>> loss(W@X)는 상수, 미분하면 사라진다
> Loss 안에 z = X@W가 있기 때문에 W+h로 변화량을 추가해서 W값을 변경하면, Loss값이 미세하게 달라진다(즉, 특정 w와 연동해서 변한다)

[4] dL / dw1 의 의미?
>w1이 h만큼 이동할 때 Loss가 수치만큼 이동한다

## 1. 미분 핵심 정리
```sage
import numpy as np

x = np.array([4. , 5.])

def function(x):
    return x[0] ** 2 + x[1] ** 2

def numerical_gradient(x, f):
    h = 1e-4
    grads = np.zeros_like(x) 
    for idx in range(x.size):
        tmp = x[idx]
        x[idx] = tmp + h
        fxh1 = f(x)
        x[idx] = tmp - h
        fxh2 = f(x)
        grads[idx] = (fxh1-fxh2)/(2*h)
        x[idx] = tmp
    return grads

def gradient_descent(f, init_x, lr=0.01, step_num=100):
    x = init_x
    for i in range(step_num):
        grads = numerical_gradient(x, f)
        x = x - lr*grads
    return x

numerical_gradient(x, function)
gradient_descent(function, x, lr=0.01, step_num=5)
```

### 편미분할 때만 에러있음
```
def function_2(x):
    return np.sum(x**2)
```

## 2. 최적화(Gradient Descent - Loss function) 
출력망에서 뱉어낸 y와 t를 이용한 손실함수를 미분한다
가중치 편미분:  dL / dW
편향 편미분: dL / db

**모든 파라미터(가중치, 편향)의 미분값이 0**이어야 최솟값의 필요조건이다
### dL / dθ = 0  → min Loss

가중치 편미분에서 각각의 element의 의미?
x축 w_ij과 y축 Loss(MSE) 간의 Convex func.에서 최솟값을 각각 구해서 행렬로 묶는다

### 핵심 : **편향이 0, 가중치 1만 아니**면 아무거나 써도 됨
```sage
class SimpleNet:
    # 신경망 내 초기 매개변수 셋팅
    def __init__(self):
        # 1) 정규분포 랜덤 * 0.01
        # 2) 카이밍 히 초기값 (He 초깃값)(ReLU와 같이씀)
        # 3) Xavier 초기값 (Sigmoid와 같이씀)
        self.W = np.random.randn(2, 3)
    
    # 예측할 때는 softmax(활성화함수) 안 써도 ㄱㅊ함
    def predict(self, x):
        return x @ self.W
    
    # loss를 구할 때 예측값과 target이 필요
    def loss(self, x, t):
        z = self.predict(x)
        y = softmax(z)
        loss = cross_entropy_error(y, t)
        return loss

net = SimpleNet()    
net.W

x = np.array([0.6, 0.9])
p = net.predict(x)

t = np.array([0, 1, 0]) # 정답을 맞춤
t_error = np.array([1, 1, 1]) # 1로만 다 추측
l = net.loss(x,t)
l_error = net.loss(x, t_error)
print(l, l_error)  # 0.3594, 1.9859
```

## Loss func. - Partial Derivative 결합
손실함수를 호출하는 즉시 손실함수는 상수가 된다
즉, 편미분하려면 함수를 편미분해야하는데 상수가 나오므로 문제 발생!!

### 즉, W는 **쓰인다!**
```sage
def f(W):
    return net.loss(x, t)

# x와 t는 따로 위에 정의했다
loss_W = lambda W : net.loss(x, t)
```

### 핵심
> 1) 미분때 벗겨질 방패막이용 f() 함수를 만든다
>> **카트라이더 실드** 느낌
> 2) loss 함수와 연동되도록 return에 net.loss를 넣는다

### w가 조절될 때 Loss가 변하는 정도를 알기 위해 경사하강법 ㄱㄱ
```sage
dW = numerical_gradient(f, net.W)
dW
```
